src/query_engine/query_element/LinkTemplate.h

```diff
@@ -0,0 +1,386 @@
+            }
+        }
+        this->handle = shared_ptr<char>(composite_hash(this->handle_keys, ARITY + 1));
+        if (! wildcard_flag) {
+            free(this->handle_keys[0]);
+        }
+        // This is correct. id is not necessarily a handle but an identifier. It just happens
+        // that we want the string for this identifier to be the same as the string representing
+        // the handle.
+        this->id = this->handle.get();
+    }
+
+    /**
+     * Destructor.
+     */
+    virtual ~LinkTemplate() {
+        graceful_shutdown();
+        local_answers_mutex.lock();
+        if (local_answers_size > 0) {
+            delete [] this->atom_document;
+            delete [] this->local_answers;
+            delete [] this->next_inner_answer;
+        }
+        local_answers_mutex.unlock();
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // QueryElement API
+
+    /**
+     * Gracefully shuts down this QueryElement's processor thread.
+     */
+    virtual void graceful_shutdown() {
+        set_flow_finished();
+        if (this->local_buffer_processor != NULL) {
+            this->local_buffer_processor->join();
+            this->local_buffer_processor = NULL;
+        }
+        Source::graceful_shutdown();
+    }
+
+    virtual void setup_buffers() {
+        Source::setup_buffers();
+        if (this->inner_template.size() > 0) {
+            switch(this->inner_template.size()) {
+                case 1: {
+                    this->inner_template_iterator = shared_ptr<Iterator>(new Iterator(
+                        inner_template[0]
+                    ));
+                    break;
+                }
+                case 2: {
+                    this->inner_template_iterator = shared_ptr<Iterator>(new Iterator(
+                        new And<2>({
+                            inner_template[0],
+                            inner_template[1]
+                        }),
+                        true
+                    ));
+                    break;
+                }
+                case 3: {
+                    this->inner_template_iterator = shared_ptr<Iterator>(new Iterator(
+                        new And<3>({
+                            inner_template[0],
+                            inner_template[1],
+                            inner_template[2]
+                        }),
+                        true
+                    ));
+                    break;
+                }
+                case 4: {
+                    this->inner_template_iterator = shared_ptr<Iterator>(new Iterator(
+                        new And<4>({
+                            inner_template[0],
+                            inner_template[1],
+                            inner_template[2],
+                            inner_template[3]
+                        }),
+                        true
+                    ));
```

Comment:
Not mandatory, just a tip. All these `shared_ptr` constructions could be shortened this way:

```cpp
                    this->inner_template_iterator = make_shared<Iterator>(
                        new And<4>({
                            inner_template[0],
                            inner_template[1],
                            inner_template[2],
                            inner_template[3]
                        }),
                        true
                    );
```

`make_shared` implicitly calls `new Iterator(...)` and passes in the arguments to it.

Reply:
OK. I just created a card to fix this: https://github.com/singnet/das-attention-broker/issues/43